crates/types/src/block_sync.rs

```diff
@@ -0,0 +1,380 @@
+    fn reorg(&self, reorg_range: RangeInclusive<u64>);
+}
+
+/// Consumer to block sync producer
+pub trait BlockSyncConsumer: Debug + Clone + Send + Sync + Unpin + 'static {
+    fn sign_off_reorg(
+        &self,
+        module: &'static str,
+        block_range: RangeInclusive<u64>,
+        waker: Option<Waker>
+    );
+    fn sign_off_on_block(&self, module: &'static str, block_number: u64, waker: Option<Waker>);
+    fn current_block_number(&self) -> u64;
+    fn has_proposal(&self) -> bool;
+    fn fetch_current_proposal(&self) -> Option<GlobalBlockState>;
+    fn register(&self, module_name: &'static str);
+
+    fn can_operate(&self) -> bool {
+        !self.has_proposal()
+    }
+}
+
+/// The global block sync is a global syncing mechanism.
+///
+/// every module that interacts with blocks and or other modules that are block
+/// sensitive are registered with the global block sync. The global block number
+/// will only increase when all modules that are registered mark the global sync
+/// as ready to increment. Unlike progressing. Any one module is able to abort
+/// the process (aborts due to reorgs)
+#[derive(Debug, Clone)]
+pub struct GlobalBlockSync {
+    /// state that we are waiting on all sign offs for
+    pending_state:      Arc<RwLock<VecDeque<GlobalBlockState>>>,
+    /// the block number
+    block_number:       Arc<AtomicU64>,
+    /// the modules with there current sign off state for the transition of
+    /// pending state -> cur state
+    registered_modules: DashMap<&'static str, SignOffState>
+}
+
+impl GlobalBlockSync {
+    pub fn new(block_number: u64) -> Self {
+        Self {
+            block_number:       Arc::new(AtomicU64::new(block_number)),
+            pending_state:      Arc::new(RwLock::new(VecDeque::with_capacity(2))),
+            registered_modules: DashMap::default()
+        }
+    }
+
+    fn proper_proposal(&self, proposal: &GlobalBlockState) -> bool {
+        self.pending_state.read().unwrap().front() == Some(proposal)
+    }
+}
+impl BlockSyncProducer for GlobalBlockSync {
+    fn new_block(&self, block_number: u64) {
+        // add to pending state. this will trigger everyone to stop and start dealing
+        // with new blocks
+        if self.block_number.load(Ordering::SeqCst) + 1 != block_number {
+            panic!("already have this block_number");
+        }
+
+        self.pending_state
+            .write()
+            .unwrap()
+            .push_back(GlobalBlockState::PendingProgression(block_number));
+    }
+
+    fn reorg(&self, reorg_range: RangeInclusive<u64>) {
+        self.pending_state
+            .write()
+            .unwrap()
+            .push_back(GlobalBlockState::PendingReorg(reorg_range));
+    }
+}
+
+impl BlockSyncConsumer for GlobalBlockSync {
+    fn sign_off_reorg(
+        &self,
+        module: &'static str,
+        block_range: RangeInclusive<u64>,
+        waker: Option<Waker>
+    ) {
+        // check to see if there is pending state
```

Comment:
There are possible race conditions in this method since we keep grabbing and dropping locks on `self.pending_state`.  It seems like it's unlikely due to the overall construction of what we're doing here but....if we're going to build locks for concurrency we should probably be thorough

Reply:
Yeah fair. One assumption I noticed and am not a fan of here is the assumption that all sign_offs for a given state happen at once before any other module tries to signoff for another state. There is a case here were lets say, reorg and new block happen very close. A module that can process and adjust to these super quick might be able to sign off on both before the first one is through causing a panic. I'm going to go through tomorrow and make it so you can sign off on multiple proposals. 

Cases thought of.
1) due to the fact that a reorg has to be registered by the producer before the consumer can react. I have made it such that the events (Cannonical State Notifications) that trigger these events are rebroadcast at the point where the producer registers them. This ensures that the proposal will be in the block sync before anyone can try to sign off on it

2) Racy conditions you mentioned. Although we drop and grab pending_state_lock. at the end of the day. the only thing that matters here is that it gets removed (passed) after a fixed N amount of people sign off on it. Assuming that this is a fixed set at runtime and no modules will be registering at random. I decided that this was safe as over a fixed set. No matter what races occur. there will always be someone last who looks at data and thus clears it. If we think this is to optimistic. we can refactor and then have us hold onto the module list such that anyone who tries to register while a signoff is occurring will have to wait. (realistically, entry queue for if we want to really be safe here as we don't want the case where a new module could register but not recieve the data for the proposal it needs to sign off on)